src/Servers/HttpSys/src/RequestProcessing/RequestContextOfT.Log.cs

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Server.HttpSys
+{
+    internal sealed partial class RequestContext<TContext>
```

Comment:
@davidfowl Realistically will there only ever be one context type used here at runtime?

I'm thinking about static properties on a generic type and property new instances being created for each generic parameter value.

Reply:
Only for value types. I don't think it'll matter here in practice but we can move it in another commit.

Reply:
Value types with generic parameters would create copies of IL, but the reference type delegates would still be copied between generic instances?

https://www.jetbrains.com/help/resharper/StaticMemberInGenericType.html

Reply:
Does this count as a member though? These are methods on a static nested class. I don't think that applies:

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQB4AVAPgYG8aGghgKH0mKBgGVsATwASMADaKIACgCUIwf2pC9DADIQA5gDpp8pStUAiACLYAbgEsAJjfUBuLQwC+PnwAHKGdHbAwYJnIkJlJDEx8dfVFGYmjxKVkFZTU0gAYGADtsfBhNXWS+H0qogE5VABIbbJU+YtLfD28K5P8evxo+mjFmADEICGFqJJSMuwgjY2MYKA1E6qEwqAYwBgBeIpgAd1Y2fK4NbpqthmB9w5P2CGAAKxgwDAuvDcEwcyyrGpvv09MB/pYcpcfH1fEA

Reply:
Oh wait, you're referring to the static delegates inside of the static class.

Reply:
```
System.NotSupportedException: Type C`1+Log[T] has a static constructor, which is not supported by SharpLab JIT decompiler.
 ```
Damn 😢 .

cc @benaadams do you happen to know?

Regardless, we can move it if it's a problem in theory, but there's a single T that hosting uses and it's currently a reference type. In general, I'd like to understand though 😄

Reply:
Sharplab doesn't like static ctors in generic types 🤕 

```cs
using System;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            var c = new C<string>();
            var b = new C<object>();
            c.SayHello();
            b.SayHello();
        }
    }

    public class C<T>
    {

        public void SayHello()
        {
            Log.SayHello("David");
        }

        private static class Log
        {
            public static readonly string Value = Guid.NewGuid().ToString();

            public static void SayHello(string name)
            {
                Console.WriteLine($"Hello {name} " + Value);
            }
        }
    }
}
```

![image](https://user-images.githubusercontent.com/303201/112772876-c123bd00-908f-11eb-8d60-c51382a7af7e.png)

Reply:
> Value types with generic parameters would create copies of IL, but the reference type delegates would still be copied between generic instances?

So `C<object>.Log._blah` and `C<string>.Log._blah` would end up being instantiated twice. Yes that makes sense and can be verified.

Reply:
@martincostello Can you send a follow up PR moving this specific log out of the generic type? It doesn't matter in practice but we can enforce this as a rule when looking at our logging codez.

Reply:
https://github.com/dotnet/aspnetcore/issues/31340

Reply:
> cc @benaadams do you happen to know?

Here is much of muchness as are likely to only have one `TContext` per application and if you go via hosting then its always the same type. 

However yes the `static`s will be unique per generic regardless of whether they are reference or value types.

Reply:
> However yes the `static`s will be unique per generic regardless of whether they are reference or value types.

*Aside: gives a really fast `IFeatureCollection.Get<T>()` style lookup; but alas it only works as a singleton lookup so can't use that for the actual `IFeatureCollection` since that's per request 😢

Reply:
> However yes the statics will be unique per generic regardless of whether they are reference or value types.

Yea, this is obvious in hindsight 😄

Reply:
> @martincostello Can you send a follow up PR moving this specific log out of the generic type? It doesn't matter in practice but we can enforce this as a rule when looking at our logging codez.

Sure, I'll push something up today.

Reply:
Someone beat me to it - #31345